source/src/plugins/Chi2Test.cc

```diff
@@ -0,0 +1,126 @@
+    class Chi2Test : public QualityTest {
+    public:
+      class Factory : public QualityTestFactory {
+      public:
+        QTestPtr createQualityTest(const std::string &qname) const override;
+      };
+
+      Chi2Test(const std::string &name);
+      ~Chi2Test() override = default;
+      StatusCode readSettings(const dqm4hep::core::TiXmlHandle xmlHandle) override;
+      void userRun(MonitorElement* monitorElement, QualityTestReport &report) override;
+
+    protected:
+      std::string m_options;
+
+    };
+
+    typedef Chi2Test::Factory Chi2TestFactory;
+
+    //-------------------------------------------------------------------------------------------------
+    //-------------------------------------------------------------------------------------------------
+
+    inline QTestPtr Chi2Test::Factory::createQualityTest(const std::string &qname) const {
+      return std::shared_ptr<QTest>(new Chi2Test(qname));
+    }
+
+    //-------------------------------------------------------------------------------------------------
+    //-------------------------------------------------------------------------------------------------
+
+    Chi2Test::Chi2Test(const std::string &qname)
+        : QualityTest("Chi2", qname),
+	  m_options()
+{
+      m_description = "Performs the chi-squared test on a TH1 and a reference, outputting the p-value. This is analogous to the Kolmogorov-Smirnov "
+	              "test but is appropriate for use on binned data like histograms, which the Kolmogorov test is not.";
+    }
+
+    //-------------------------------------------------------------------------------------------------
+
+    StatusCode Chi2Test::readSettings(const TiXmlHandle xmlHandle) {
+      RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::readParameter(xmlHandle, "Options", m_options));
+
+      return STATUS_CODE_SUCCESS;
+    }
+
+    void Chi2Test::userRun(MonitorElement* pMonitorElement, QualityTestReport &report) {
+      const bool hasObject = (pMonitorElement->object() != nullptr);
+      const bool hasReference = (pMonitorElement->reference() != nullptr);
+      const bool isHistogram = pMonitorElement->object()->InheritsFrom("TH1");
+
+      Option_t* pOptions = &m_options[0u];
+
+      if (!hasObject) {
+        report.m_message = "No object attached to monitor element";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      if (!isHistogram) {
+        report.m_message = "The object must be a TH1";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      if (!hasReference) {
+        report.m_message = "No reference attached to monitor element";
+        report.m_quality = 0.f;
+        throw StatusCodeException(STATUS_CODE_INVALID_PTR);
+      }
+
+      TH1* pHistogram = pMonitorElement->objectTo<TH1>();
+      TH1* pReferenceHistogram = pMonitorElement->referenceTo<TH1>();
```

Comment:
The framework doesn't ensure that if the object is TH1 then the reference is also TH1. It can be any other type. You have to test inheritance here also.